appsec-kit-v24/src/main/java/com/vaadin/appsec/v24/views/AppSecView.java

```diff
@@ -0,0 +1,152 @@
+        headerBar.add(buildLastScannedLabel());
+        headerBar.add(buildScanNowButton());
+
+        getMainContent().add(headerBar);
+        getMainContent().addAndExpand(buildTabSheet());
+    }
+
+    private Component buildAppTitle() {
+        Span appTitle = new Span("AppSec Kit");
+        appTitle.addClassName("appsec-kit-title");
+        appTitle.setSizeFull();
+        return appTitle;
+    }
+
+    private Component buildLastScannedLabel() {
+        lastScannedLabel = new Span();
+        lastScannedLabel.setWidth(380, Unit.PIXELS);
+        return lastScannedLabel;
+    }
+
+    private Component buildScanNowButton() {
+        scanNowButton = new Button("Scan now");
+        scanNowButton.setDisableOnClick(true);
+        scanNowButton.addClickListener(e -> {
+            lastScannedLabel.setText("Scanning...");
+            AppSecService.getInstance().scanForVulnerabilities();
+        });
+        return scanNowButton;
+    }
+
+    private Component buildTabSheet() {
+        tabSheet = new TabSheet();
+        tabSheet.setSizeFull();
+        tabSheet.addSelectedChangeListener(e -> {
+            Component component = tabSheet
+                    .getComponent(tabSheet.getSelectedTab());
+            if (component instanceof AbstractAppSecView abstractAppSecView) {
+                abstractAppSecView.refresh();
+            }
+        });
+        vulnerabilitiesTab = new VulnerabilitiesTab(this);
+        dependenciesTab = new DependenciesTab(this);
+        tabSheet.add("Vulnerabilities", vulnerabilitiesTab);
+        tabSheet.add("Dependencies", dependenciesTab);
+        return tabSheet;
+    }
+
+    @Override
+    public void refresh() {
+        Component component = tabSheet.getComponent(tabSheet.getSelectedTab());
+        if (component instanceof AbstractAppSecView abstractAppSecView) {
+            abstractAppSecView.refresh();
+        }
+        Instant lastScan = AppSecService.getInstance().refresh().getLastScan();
+        lastScannedLabel.setText("Last Scan: " + (lastScan == null ? "--"
+                : formatter.format(Date.from(lastScan))));
+    }
+
+    void showVulnerabilitiesTabFor(Dependency item) {
+        tabSheet.setSelectedTab(tabSheet.getTab(vulnerabilitiesTab));
+        vulnerabilitiesTab.filterOn(item);
+    }
+
+    @Override
+    public void onAttach(AttachEvent event) {
+        super.onAttach(event);
+        removeScanListener();
+        scanListener = AppSecService.getInstance()
+                .addScanEventListener(scanEvent -> AppSecView.this.getUI()
+                        .ifPresent(ui -> ui.access(() -> {
+                            scanNowButton.setEnabled(true);
+                            refresh();
+                            if (PushMode.MANUAL == ui.getPushConfiguration()
+                                    .getPushMode()) {
+                                ui.push();
+                            }
+                        })));
```

Comment:
Both the scan event-listener and the UI-access command could be extracted to methods to make this more readable.

Reply:
Refactored this method. Also refactored the same solution in V7 and V8 to be consistent.